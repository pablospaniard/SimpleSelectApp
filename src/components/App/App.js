import React, { Component, Fragment } from 'react'
import axios from 'axios'
import { Button, Paper, Grid, Typography } from '@material-ui/core'
import styled from 'styled-components'

import { Applied, Interviewing, Hired } from '../'
import { Spinner } from '../UI'

const StyledGrid = styled(Grid)`
  display: flex;
  align-items: middle;
  justify-content: center;
`

class App extends Component {
  state = {
    persons: {},
    step: 0,
    loading: true,
    selected: [],
    hired: []
  }

  componentDidMount = () => {
    axios
      .get(' https://randomuser.me/api/?nat=gb&results=5')
      .then(res => {
        this.setState({ persons: res.data, loading: false })
      })
      .catch(err => console.log(err))
  }

  onSelectandler = email => {
    const { selected } = this.state
    if (selected.indexOf(email) === -1) {
      const newSelected = selected
      newSelected.push(email)
      this.setState({ selected: newSelected })
    } else {
      const newSelected = selected
      const index = selected.indexOf(email)
      newSelected.splice(index, 1)
      this.setState({ selected: newSelected })
    }
  }

  onHiredHandler = email => {
    const { hired } = this.state
    if (hired.indexOf(email) === -1) {
      const newHired = hired
      newHired.push(email)
      this.setState({ hired: newHired })
    } else {
      const newHired = hired
      const index = hired.indexOf(email)
      newHired.splice(index, 1)
      this.setState({ hired: newHired })
    }
  }

  onNext = () => {
    const { step } = this.state
    this.setState({ step: step + 1 })
  }

  onPreviuos = () => {
    const { step } = this.state
    if (step === 1) {
      this.setState({ step: step - 1, selected: [] })
    }
    if (step === 2) {
      this.setState({ step: step - 1, hired: [] })
    }
  }

  render() {
    const { step, persons, loading, selected, hired } = this.state
    let content

    if (loading) {
      content = <Spinner />
    }

    if (!loading) {
      switch (step) {
        case 0:
          content = (
            <Fragment>
              <Typography variant="headline" style={{ padding: '20px' }}>
                Applied
              </Typography>
              <Applied
                persons={persons.results}
                onChangeHandler={this.onSelectandler}
              />
            </Fragment>
          )
          break

        case 1:
          content = (
            <Fragment>
              <Typography variant="headline" style={{ padding: '20px' }}>
                Interviewing
              </Typography>
              <Interviewing
                persons={persons.results}
                onChangeHandler={this.onHiredHandler}
                selected={selected}
              />
            </Fragment>
          )
          break

        case 2:
          content = (
            <Fragment>
              <Typography variant="headline" style={{ padding: '20px' }}>
                Hired
              </Typography>
              <Hired persons={persons.results} hired={hired} />
            </Fragment>
          )
          break

        default:
          break
      }
    }

    return (
      <Grid container>
        <StyledGrid item xs={2}>
          {this.state.step === 0 ? null : (
            <Button color="primary" onClick={this.onPreviuos}>
              Previuos
            </Button>
          )}
        </StyledGrid>
        <Grid item xs={8}>
          <Paper>{content}</Paper>
        </Grid>
        <StyledGrid item xs={2}>
          {this.state.step === 2 ||
          (selected.length === 0 && this.state.step === 1) ? null : (
            <Button color="secondary" onClick={this.onNext}>
              Next
            </Button>
          )}
        </StyledGrid>
      </Grid>
    )
  }
}

export default App
