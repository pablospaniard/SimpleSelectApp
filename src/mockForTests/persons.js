export const persons = [
  {
    gender: 'female',
    name: {
      title: 'ms',
      first: 'nicky',
      last: 'vargas'
    },
    location: {
      street: '9761 station road',
      city: 'salisbury',
      state: 'cumbria',
      postcode: 'NN4 9LF',
      coordinates: {
        latitude: '-5.3269',
        longitude: '48.1334'
      },
      timezone: {
        offset: '+2:00',
        description: 'Kaliningrad, South Africa'
      }
    },
    email: 'nicky.vargas@example.com',
    login: {
      uuid: 'bee8fcb3-1261-47ff-ab6f-75c7be01077c',
      username: 'yellowladybug417',
      password: 'gogogo',
      salt: 'ixgD083Q',
      md5: '3c6400fffd6adddba4df6856079d61fd',
      sha1: '8f109908f4db48a9a8743ecf1ecc0d66e2536888',
      sha256: 'be9fd8363e25c08f58fb111600673a3c548b9f158fc847b5c48212f5fb2fa3c3'
    },
    dob: {
      date: '1986-01-13T11:28:23Z',
      age: 32
    },
    registered: {
      date: '2009-09-11T00:05:01Z',
      age: 9
    },
    phone: '016973 35595',
    cell: '0707-619-299',
    id: {
      name: 'NINO',
      value: 'TZ 01 18 38 X'
    },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/69.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/69.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/69.jpg'
    },
    nat: 'GB'
  }
]
