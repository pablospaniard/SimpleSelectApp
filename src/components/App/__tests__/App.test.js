import React from 'react'
import {
  render,
  cleanup,
  waitForElement,
  fireEvent
} from 'react-testing-library'

import { persons } from '../../../mockForTests/persons'
import App from '../App'

describe('App should render properly', () => {
  afterEach(cleanup)

  it('should match snapshot', () => {
    const { container } = render(<App />)
    expect(container.firstChild).toMatchSnapshot('App_snapshot_1')
  })

  it('render Applied component if step is 0', async () => {
    const { getByText } = render(<App />)
    await waitForElement(() => getByText('Applied'))
  })

  it('render Interviewing component if next button is pressed', async () => {
    const { getByText } = render(<App />)
    fireEvent.click(getByText('Next'))
    await waitForElement(() => getByText('Interviewing'))
  })
})
