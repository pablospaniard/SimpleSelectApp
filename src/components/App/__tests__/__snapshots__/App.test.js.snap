// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App_snapshot_1 1`] = `
<div
  class="MuiGrid-container-1"
>
  <div
    class="MuiGrid-item-2 MuiGrid-grid-xs-2-31 sc-bwzfXH fUBRka"
  />
  <div
    class="MuiGrid-item-2 MuiGrid-grid-xs-8-37"
  >
    <div
      class="MuiPaper-root-98 MuiPaper-elevation2-102 MuiPaper-rounded-99"
    >
      <div
        class="sc-bdVaJa gDMQUI"
        data-testid="spinner"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-item-2 MuiGrid-grid-xs-2-31 sc-bwzfXH fUBRka"
  >
    <button
      tabindex="0"
      class="MuiButtonBase-root-151 MuiButton-root-125 MuiButton-text-127 MuiButton-textSecondary-129 MuiButton-flat-130 MuiButton-flatSecondary-132"
      type="button"
    >
      <span
        class="MuiButton-label-126"
      >
        Next
      </span>
      <span
        class="MuiTouchRipple-root-154"
      />
    </button>
  </div>
</div>
`;
